src/main/rascal/examples/pico/Rename.rsc

```diff
@@ -41,93 +41,46 @@ import util::FileSystem;
 
 data Tree;
 
-alias RenameRequest = tuple[list[Tree] cursor, str newName/*, set[loc] workspaceFolders*/];
-
-data RenameState
-    = findDefinition(RenameRequest req)
-    | rename(IdRole role, loc def, str newName)
-    ;
-
-@memo{expireAfter(minutes=1), maximumSize(50)}
-Tree parseLoc(loc l) {
-    return parse(#start[Program], l);
-}
-
-@memo{expireAfter(minutes=1), maximumSize(50)}
-TModel tmodelForLoc(loc l) {
-    return collectAndSolve(parseLoc(l));
-}
+public tuple[list[DocumentEdit] edits, map[str, ChangeAnnotation] annos, set[Message] msgs] renamePico(list[Tree] cursor, str newName) {
+    if (!isValidName(newName)) {
+        return <[], (), {error("\'<newName>\' is not a valid name here.", cursor[0].src)}>;
+    }
 
-public tuple[list[DocumentEdit] edits, map[str, ChangeAnnotation] annos, set[Message] msgs] renamePico(RenameRequest request) {
-    RenameConfig config = rconfig(
-        parseLoc
-      , tmodelForLoc
-      , reportCollectCycles = true
+    return rename(
+        cursor
+      , newName
+      , Tree(loc l) { return parse(#start[Program], l); }
+      , collectAndSolve
+      , findDefinitions
+      , findCandidateFiles
+      , renameDef
     );
-
-    RenameSolver solver = newSolverForConfig(config);
-    initSolver(solver, request);
-    return solver.run();
 }
 
-void initSolver(RenameSolver solver, RenameRequest req) {
-    if (!isValidName(req.newName)) {
-        // gooi errors
-        solver.msg(error("Not a valid name: \'<req.newName>\'!"));
-        return;
+set[Define] findDefinitions(list[Tree] cursor, TModel(Tree) getTModel, Renamer r) {
+    TModel tm = getTModel(cursor[-1]);
+    if (Tree t <- cursor
+      , tm.definitions[t.src]?) {
+            return {tm.definitions[t.src]};
```

Comment:
shouldn't this be a set already?

Reply:
No, `tm.definitions` is a `map[loc, Define]`